pkg/operator/etcdcertcleaner/etcd_cert_cleaner_controller.go

```diff
@@ -0,0 +1,187 @@
+
+import (
+	"context"
+	"fmt"
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/ceohelpers"
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/health"
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/operatorclient"
+	"github.com/openshift/cluster-etcd-operator/pkg/tlshelpers"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+	"strings"
+	"time"
+)
+
+const DeletionGracePeriodAnnotation = "openshift.io/ceo-delete-grace-start"
+const DeletionGracePeriod = 24 * 3 * time.Hour
+
+var dynamicCertPrefixes = []string{
+	tlshelpers.GetPeerClientSecretNameForNode(""),
+	tlshelpers.GetServingSecretNameForNode(""),
+	tlshelpers.GetServingMetricsSecretNameForNode(""),
+}
+
+// EtcdCertCleanerController will clean up old and unused dynamically generated ("node-based") secrets.
+// This is to avoid filling up etcd in environments where the control plane is rotated/scaled up and down very often.
+type EtcdCertCleanerController struct {
+	eventRecorder      events.Recorder
+	kubeClient         kubernetes.Interface
+	operatorClient     v1helpers.StaticPodOperatorClient
+	masterNodeLister   corev1listers.NodeLister
+	masterNodeSelector labels.Selector
+	secretInformer     corev1informers.SecretInformer
+	secretLister       corev1listers.SecretLister
+	secretClient       corev1client.SecretsGetter
+	quorumChecker      ceohelpers.QuorumChecker
+}
+
+func NewEtcdCertCleanerController(
+	livenessChecker *health.MultiAlivenessChecker,
+	kubeClient kubernetes.Interface,
+	operatorClient v1helpers.StaticPodOperatorClient,
+	kubeInformers v1helpers.KubeInformersForNamespaces,
+	masterNodeLister corev1listers.NodeLister,
+	masterNodeSelector labels.Selector,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	secretInformer := kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Secrets()
+	secretLister := secretInformer.Lister()
+	secretClient := v1helpers.CachedSecretGetter(kubeClient.CoreV1(), kubeInformers)
+
+	c := &EtcdCertCleanerController{
+		eventRecorder:      eventRecorder,
+		kubeClient:         kubeClient,
+		operatorClient:     operatorClient,
+		masterNodeLister:   masterNodeLister,
+		masterNodeSelector: masterNodeSelector,
+		secretInformer:     secretInformer,
+		secretLister:       secretLister,
+		secretClient:       secretClient,
+	}
+
+	// we estimate the rate of control plane scales to maybe 1-2 a day, so running this once an hour with 6h of health threshold
+	// should be more than sufficient. Note that this controller does not sync on any informers.
```

Comment:
Sounds fair. Even if there's an abnormal amount of scaling and we have a pile up of unused secrets and you need to clean them before the hour mark, you can always manually clean them. Or just restart the operator to resync the controller.

Reply:
yep, after our v-scaling work I'd have some trouble finding a hyperscaler that could scale that fast :)